src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,497 @@
+                entry.marker << " (" << entry_idx << "/" << total_entries << ") from: " << queue_name << " ok" << dendl;
+              remove_entries = true;
+            }  else {
+              if (set_min_marker(end_marker, entry.marker) < 0) {
+                ldout(cct, 1) << "ERROR: cannot determin minimum between malformed markers: " << end_marker << ", " << entry.marker << dendl;
+              } else {
+                ldout(cct, 20) << "INFO: new end marker for removal: " << end_marker << " from: " << queue_name << dendl;
+              }
+              has_error = true;
+              ldout(cct, 20) << "INFO: processing of entry: " << 
+                entry.marker << " (" << entry_idx << "/" << total_entries << ") from: " << queue_name << " failed" << dendl;
+            } 
+        });
+        ++entry_idx;
+      }
+
+      // wait for all pending work to finish
+      waiter.async_wait(yield);
+
+      // delete all published entries from queue
+      if (remove_entries) {
+        librados::ObjectWriteOperation delete_op;
+        cls_2pc_queue_remove_entries(delete_op, end_marker); 
+        const auto ret = rgw_rados_operate(rados_ioctx, queue_name, &delete_op, optional_yield(io_context, yield)); 
+        if (ret < 0) {
+          ldout(cct, 1) << "ERROR: failed to remove entries up to: " << end_marker <<  " from queue: " 
+            << queue_name << ". error: " << ret << dendl;
+        } else {
+          ldout(cct, 20) << "INFO: removed entries up to: " << end_marker <<  " from queue: " 
+          << queue_name << dendl;
+        }
+      }
+      // TODO: cleanup expired reservations
+    }
+  }
+
+  // process all queues
+  // find which of the queues is owned by this daemon and process it
+  void process_queues(spawn::yield_context yield) {
+    auto has_error = false;
+    std::unordered_set<std::string> owned_queues;
+    while (true) {
+      Timer timer(io_context);
+      const auto duration = has_error ? 
+        std::chrono::milliseconds(queues_update_retry_ms) : std::chrono::milliseconds(queues_update_period_ms);
+      timer.expires_from_now(duration);
+      const auto tp = ceph::coarse_real_time::clock::to_time_t(ceph::coarse_real_time::clock::now() + duration);
+      ldout(cct, 20) << "INFO: next queue processing will happen at: " << std::ctime(&tp)  << dendl;
+      boost::system::error_code ec;
+      timer.async_wait(yield[ec]);
+
+      queues_t queues;
+      auto ret = read_queue_list(queues, optional_yield(io_context, yield));
+      if (ret < 0) {
+        has_error = true;
+        continue;
+      }
+
+      std::vector<std::string> queue_gc;
+      std::mutex queue_gc_lock;
+      for (const auto& queue_name : queues) {
+        // try to lock the queue to check if it is owned by this rgw
+        // or if ownershif needs to be taken
+        librados::ObjectWriteOperation op;
+        rados::cls::lock::lock(&op, queue_name+"_lock", 
+              LOCK_EXCLUSIVE,
```

Comment:
note that this is a write operation, so it will create the object if it doesn't exist. it looks like this could race with `remove_persistent_topic()` and recreate the object. if that happens, you wouldn't be able to recreate the topic because `add_persistent_topic()` is doing an exclusive create

you can add a call to `op.assert_exists()` to avoid recreating the object here - if it doesn't exist, the op will fail with ENOENT